norddrop/src/device.rs

```diff
@@ -522,22 +522,29 @@ impl NordDropFFI {
     }
 }
 
-fn crate_key_context(
-    logger: slog::Logger,
-    privkey: SecretKey,
-    pubkey_cb: impl Fn(IpAddr) -> Option<PublicKey> + Send + 'static,
-) -> auth::Context {
-    let pubkey_cb = std::sync::Mutex::new(pubkey_cb);
-    let public = move |ip: IpAddr| {
-        let guard = pubkey_cb.lock().expect("Failed to lock pubkey callback");
-        let key = guard(ip)?;
-        drop(guard);
+fn crate_key_context(logger: slog::Logger, key_store: Arc<Box<dyn KeyStore>>) -> auth::Context {
+    let priv_keystore = Arc::clone(&key_store);
+
+    let priv_logger = logger.clone();
+    let privkey = move || {
+        let key_store = Arc::clone(&priv_keystore);
+
+        let privkey: [u8; PUBLIC_KEY_LENGTH] = key_store.privkey().try_into().ok()?;
+        debug!(priv_logger, "Retrieved private key: {:?}", privkey);
+        Some(SecretKey::from(privkey))
+    };
+
+    let pub_keystore = Arc::clone(&key_store);
+    let pubkey = move |ip: IpAddr| {
+        let key_store = Arc::clone(&pub_keystore);
```

Comment:
```suggestion
    let priv_keystore = key_store.clone();
    let priv_logger = logger.clone();
    let privkey = move || {
        let key_store = priv_keystore.clone();

        let privkey: [u8; PUBLIC_KEY_LENGTH] = key_store.privkey().try_into().ok()?;
        debug!(priv_logger, "Retrieved private key: {:?}", privkey);
        Some(SecretKey::from(privkey))
    };

    let pubkey = move |ip: IpAddr| {
        let key_store = key_store.clone();
```

Can get rid of one clone here, and less clutter with using straight `.clone()`